### PR TITLE
Auto Voting System

### PR DESCRIPTION
**Pull request by sleepyyapril**
# Description

Contains a single system, under the server-sided AutoVote folder, that handles automatic voting of maps and presets, both of which are configurable via a CVar.

By automatic voting system, I mean that it creates a vote for map and preset the moment players return to lobby. If the server just started and there are no players, it will run the vote when a player joins to lobby.

As of now, the best use case for this is for your players to play the modes they want, at all times.

Relevant CVars:
`vote.autovote_enabled - Is the autovote system enabled at all? (Default is false)`
`vote.map_autovote_enabled - If it is applicable, is automatic map voting enabled? (Default is true)`
`vote.preset_autovote_enabled - If it is applicable, is automatic preset voting enabled? (Default is true)`

---

# TODO

- [x] Base system
- [x] CVars
- [x] Set default CVar value to false, as it is not something you might want by default. 

---

# Changelog

:cl:
- add: Added the automatic voting system.

### CODE REVIEW

File: Content.Server/AutoVote/AutoVoteSystem.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        if (_playerManager.PlayerCount == 0)
+        {
+            _shouldVoteNextJoin = true;
+            return;
+        }
+
+        if (_cfg.GetCVar(CCVars.MapAutoVoteEnabled))
+            _voteManager.CreateStandardVote(null, StandardVoteType.Preset);
+        if (_cfg.GetCVar(CCVars.PresetAutoVoteEnabled))
+            _voteManager.CreateStandardVote(null, StandardVoteType.Map);
```
Comment (FoxxoTrystan) [inline Content.Server/AutoVote/AutoVoteSystem.cs:55]:
```suggestion
        if (_cfg.GetCVar(CCVars.MapAutoVoteEnabled))
            _voteManager.CreateStandardVote(null, StandardVoteType.Map);
        if (_cfg.GetCVar(CCVars.PresetAutoVoteEnabled))
            _voteManager.CreateStandardVote(null, StandardVoteType.Preset);
```
Ehh should those not be inverted?

**Review by FoxxoTrystan — DISMISSED** [review]
I have a suggestion, as a CVar, make a vote for calling the crew exchange shuttle.

Near the end of the 2 hours mark.
If you dont wana do in this PR il still approve it as its looks nice and il add it personnaly later.

File: Content.Server/AutoVote/AutoVoteSystem.cs (hunk 1 above)
Comment (sleepyyapril) [reply Content.Server/AutoVote/AutoVoteSystem.cs:55]: Doesn't realistically make a difference as there's no return, honestly.

**Comment by sleepyyapril** [comment]
> I have a suggestion, as a CVar, make a vote for calling the crew exchange shuttle.
> 
> Near the end of the 2 hours mark. If you dont wana do in this PR il still approve it as its looks nice and il add it personnaly later.

Likely going to do this in a separate PR, with a start at and interval cvar.

**Comment by DEATHB4DEFEAT** [comment]
> > I have a suggestion, as a CVar, make a vote for calling the crew exchange shuttle.
> > Near the end of the 2 hours mark. If you dont wana do in this PR il still approve it as its looks nice and il add it personnaly later.
> 
> Likely going to do this in a separate PR, with a start at and interval cvar.

Use some of the existing autocall stuff.

File: Content.Server/AutoVote/AutoVoteSystem.cs (hunk 1 above)
Comment (DEATHB4DEFEAT) [reply Content.Server/AutoVote/AutoVoteSystem.cs:55]: To match the cvardefs

File: Content.Shared/CCVar/CCVars.cs [hunk 2]
```diff
@@ -2640,5 +2640,27 @@ public static readonly CVarDef<float>
...
+
+        /// <summary>
+        ///     Automatically make map votes on return to lobby? Requires auto voting to be enabled.
+        /// </summary>      
+        public static readonly CVarDef<bool> MapAutoVoteEnabled =
+            CVarDef.Create("vote.map_autovote_enabled", true, CVar.SERVERONLY);
+
+        /// <summary>
+        ///     Automatically make preset votes on return to lobby? Requires auto voting to be enabled.
+        /// </summary>
```
Comment (DEATHB4DEFEAT) [inline Content.Shared/CCVar/CCVars.cs:2662]:
```suggestion
        /// Enables the automatic voting system.
        public static readonly CVarDef<bool> AutoVoteEnabled =
            CVarDef.Create("vote.autovote_enabled", false, CVar.SERVERONLY);

        /// Automatically starts a map vote when returning to the lobby.
        /// Requires auto voting to be enabled.  
        public static readonly CVarDef<bool> MapAutoVoteEnabled =
            CVarDef.Create("vote.map_autovote_enabled", true, CVar.SERVERONLY);

        /// Automatically starts a gamemode vote when returning to the lobby.
        /// Requires auto voting to be enabled.  
```

**Review by DEATHB4DEFEAT — APPROVED** [review]

File: Content.Server/AutoVote/AutoVoteSystem.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        CallAutovote();
+    }
+
+    public void OnPlayerJoinedLobby(PlayerJoinedLobbyEvent ev)
+    {
+        if (_shouldVoteNextJoin)
+        {
+            CallAutovote();
+            _shouldVoteNextJoin = false;
+        }
```
Comment (VMSolidus) [inline Content.Server/AutoVote/AutoVoteSystem.cs:38]:
```suggestion
        if (!_shouldVoteNextJoin)
            return;

        CallAutovote();
        _shouldVoteNextJoin = false;
```

**Review by VMSolidus — CHANGES_REQUESTED** [review]

File: Content.Server/AutoVote/AutoVoteSystem.cs (hunk 3 above)
Comment (DEATHB4DEFEAT) [reply Content.Server/AutoVote/AutoVoteSystem.cs:38]: now we have to wait another 5 minutes for tests!

**Comment by sleepyyapril** [comment]
> > > I have a suggestion, as a CVar, make a vote for calling the crew exchange shuttle.
> > > Near the end of the 2 hours mark. If you dont wana do in this PR il still approve it as its looks nice and il add it personnaly later.
> > 
> > 
> > Likely going to do this in a separate PR, with a start at and interval cvar.
> 
> Use some of the existing autocall stuff.

The idea I had was to call a vote whenever the automatic shuttle call was sent.

File: Content.Server/AutoVote/AutoVoteSystem.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        base.Initialize();
+
+        SubscribeLocalEvent<RoundRestartCleanupEvent>(OnReturnedToLobby);
+        SubscribeLocalEvent<PlayerJoinedLobbyEvent>(OnPlayerJoinedLobby);
+    }
+
+    public void OnReturnedToLobby(RoundRestartCleanupEvent ev)
+    {
+        CallAutovote();
+    }
```
Comment (VMSolidus) [inline Content.Server/AutoVote/AutoVoteSystem.cs:30]:
```suggestion
    public void OnReturnedToLobby(RoundRestartCleanupEvent ev) => CallAutovote();
```

**Review by VMSolidus — CHANGES_REQUESTED** [review]
I googled "c# methods for getting rid of curly braces"
And now I have a funny new tool.

**Review by VMSolidus — APPROVED** [review]

**Pull request merged**